ESPS/general/src/tofspec/tofspec.1
.\" Copyright (c) 1988-90 Entropic Speech, Inc.
.\" Copyright (c) 1991-93 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)tofspec.1	1.10 3/31/97 ESI/ERL
.ds ]W (c) 1993 Entropic Research Laboratory, Inc.
.if n .ds - ---
.if t .ds - \(em\h'-0.2m'\(em
.TH TOFSPEC 1\-ESPS 3/31/97
.SH "NAME"
tofspec \- transforms data from arbitrary FEA field to FEA_SPEC file
.SH "SYNOPSIS"
.B tofspec
[
.BI \-d
] [
.BI \-f " fea_field"
] [
.BI \-i " input_range"
] [
.BI \-o " output_range"
] [
.BI \-r " record_range"
] [
.BI \-s " sf"
] [
.BI \-v " freqs"
] [
.BI \-x " debug_level"
] [
.BI \-F " freq_format"
] [
.BI \-P " params"
] [
.BI \-R
] [
.BI \-S
]
.I "input output"
.SH "DESCRIPTION"
.PP
.I Tofspec
accepts an arbitrary FEA file \fIinput\fP and produces a FEA_SPEC file
\fIoutput\fP.  The "power spectrum" in the FEA_SPEC output \*- i.e.,
the contents of the re_spec_val field \*- is taken from the field in
\fIinput\fP specified by the \fB\-f\fP option.  The main purpose of
\fItofspec\fP is to allow \fIwaves\fP+ users to display arbitrary FEA
data in spectrogram form.
.PP
The output FEA_SPEC file has frequency format SYM_EDGE by default (see
.IR FEA_SPEC (5\-ESPS)).
However, there is an option for specifying frequency format ARB_FIXED instead,
which also requires specifying an explicit list of frequencies to be stored
in the output file header (see the
.B \-F
and
.B \-v
options).
The spectrum type is DB (logarithmic power spectral density in decibels; see
.IR FEA_SPEC (5\-ESPS)).
To conserve space, the file is stored in BYTE format.
Unless the\fB\-S\fP option (no scaling) is
specified, \fITofspec\fP clips the input data to the input range
specified by the \fB\-i\fP option, scales the resulting data to the
default output range (\-64, +50), and stores the scaled data as float
values in the re_spec_val field of \fIoutput\fP.  The output range can
be changed from the default range of (\-64, +50) by means of the
\fB\-o\fP option, although the BYTE format of the output file forces
the range to stay within (\-64, +63).  If the \fB\-i\fP option is not
used, the input data are not clipped \*- the miminum and maximum values
are determined from the input data and this range is scaled to the
output range.  If the \fB\-S\fP option is specified, no scaling is
performed \*- input data are copied directly to output data with
clipping at the limits of the output range.  If the \fB\-d\fP is
specified, the log (base 10) of the (possibly clipped) input data is
taken before the data are stored (after possible scaling) in
re_spec_val of \fIoutput\fP.
.PP
The data type of the field \fIfea_field\fP can be any supported FEA
data type.  If the type is complex, only the real part is used.  
.PP
If \fIinput\fP is "\-" and the input range is specified by the
\fB\-i\fP option or is determined from the parameter file, then the
input is taken from standard input. (Standard input can not be used if
the input range is determined automatically from the input data.)  If
\fIoutput\fP is "\-", the output is directed to standard output.
.PP
The default output range provides 115 levels, from \-64 to +50.  This
is intended to use the first 115 entries in the \fIwaves\fP+ color
map.  The entire color map has 128 values, but the top 13 levels are
used for cursors, borders, and backgrounds.  Thus, if the full 128
levels are covered in the \fItofspec\fP output, various colors will
intrude when data is displayed using the grey scale map. 
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-d 
Apply log (base 10) to input data before scaling (no \fB\-S\fP) or 
storing (\fB\-S\fP) .  
.TP
.BI \-f " field_name" "\fR [sd]"
Specifies the name of the data field in \fIinput\fP to be converted.
The default name is "sd".
.TP
.BI \-i " low_input" : "high_input" " \fR[(determined from data)]"
.TP
.BI \-i " low_input" :+ "incr"
The first form gives the low and high limits of the range of input
data to scale to the output range.  This option is ignored if the
\fB\-S\fP option is used.  
If 
.IR high_input " = " low_input " + " incr,
the second form (with the plus sign) specifies the same range as the
first.  If this option is not used, the limits are determined from the
data.  
.TP
.BI \-o " low_output" : "high_output" " \fR[\-64:+50]"
.TP
.BI \-o " low_output" :+ "incr"
The first form gives the low and high limits of the output 
range.  
If 
.IR high_output " = " low_output " + " incr,
the second form (with the plus sign) specifies the same range as the
first.  If \fIlow_output\fP < \-64, it is reset to \-64 and a warning is
printed.  If \fIhigh_output\fP > 63, it is set to 63 and a warning is
printed.  This option specifies the range of output data.  Unless
the\fB\-S\fP option is specified, the input range is scaled to the
output range (see also \fB\-d\fP).  If \fB\-S\fP is specified, the input
range is ignored, and the input data are clipped to the output 
range (see also \fB\-d\fP). 
.TP
.BI \-r " start" : "last" "\fR [1:(last in file)]"
.TP
.BI \-r " start" :+ "nan"
In the first form, a pair of unsigned integers specifies the range
of records to be processed.
Either
.I start
or
.I last
may be omitted; then the default value is used.
If
.IR last " = " start " + " nan,
the second form (with the plus sign) specifies the same range as the first.
The \fB\-r\fP overrides the values of \fIstart\fP and \fInan\fP from 
the parameter file.  
.TP
.BI \-s " sf" "\fR [0]"
Specifies the value to be used for the generic header item \fIsf\fP
in \fIoutput\fP.
As a result, if the frequency format of
.I output
is SYM_CTR, the frequency scale will range from 0
to \fIsf\fP/2 when \fIoutput\fP is displayed with \fIwaves\fP+.
(If the output
.I freq_format
is ARB_FIXED,
.I sf
plays no role in setting the frequency scale, which is determined
by an explicit list of values in the header instead\*-see the
.B \-F
and
.B \-v
options.)
If the option argument is 0 (the default), the generic's value is set to
2*(\fIsize\fP \- 1), where \fIsize\fP is the size of the input field
\fIfea_field\fP (this is done so that the implicit "frequencies"
correspond to element number in the input field).
.TP
.BI \-v " freqs" "\fR [(none)]"
A list of frequencies to be stored as the value of the generic header item
.I freqs
in the output file (see
.IR FEA_SPEC (5\-ESPS)).
This option is ignored unless the
.I freq_format
of the output file is specified as ARB_FIXED, either with the
.B \-F
option or with the
.I freq_format
parameter in a parameter file.
The format of the argument is a list of real numbers,
separated by commas or blanks.
If blanks are present, they must be escaped with back slashes (\\),
or the argument must be enclosed in double quotes (").
.TP
.BI \-x " debug_level" "\fR [0]"
Print diagnostic messages as program runs (for debugging purposes
only).
Higher levels give more messages.
The default level of zero suppresses all debugging messages.
.TP
.BI \-F " freq_format" "\fR [SYM_EDGE]"
The value to assign to the
.I freq_format
generic header item in the output file (see
.IR FEA_SPEC (5\-ESPS)).
Currently supported values are SYM_EDGE and ARB_FIXED (case-insensitive).
If ARB_FIXED is specified,
a list of frequency values must be given also, either with the
.B \-v
option or with the
.I freqs
parameter in a parameter file.
.TP
.BI \-P " params" " \fR[params]\fP"
Specifies the name of the parameter file.
.TP
.BI \-R
Reverse the order of elements in the \fIfea_field\fP before writing 
to \fIoutput\fP.  This is useful if you want to invert the orientation
of the display shown when \fIoutput\fP is displayed with \fIwaves\fP+.  
.TP
.BI \-S
Do not scale the data.  This means that any input range is ignored.  
The input data (or the log of the input data, if \fB\-d\fP) 
are stored in \fIoutput\fP with clipping at the output range limits. 
.SH "ESPS PARAMETERS"
.PP
The parameter file does not have to be present, since all the
parameters have default values.  The following parameters are read, if
present, from the parameter file:
.TP
.I "field_name \- string"
This is the name of the source field in \fIinput\fP.
The default is "sd".  This parameter is not read if the \fB\-f\fP
option is used.  
.TP
.I "freq_format \- string"
The value to assign to the
.I freq_format
generic header item in the output file (see
.IR FEA_SPEC (5\-ESPS)).
This parameter is not read if the
.B \-F
command-line option is specified.
Currently supported values are SYM_EDGE (the default) and ARB_FIXED
(both case-insensitive).
If ARB_FIXED is specified,
a list of frequency values must be given with the
.B \-v
option or the
.I freqs
parameter.
.TP
.I "freqs \- float array"
A list of frequencies to be stored as the value of the generic header item
.I freqs
in the output file (see
.IR FEA_SPEC (5\-ESPS)).
This option is not read if the
.B \-v
command-line option is specified,
and it is ignored unless the
.I freq_format
of the output file is specified as ARB_FIXED with the
.B \-F
option or the
.I freq_format
parameter.
.TP
.I "start \- integer"
This is the first record of \fIinput\fP to process.  The default is 1.  It
is not read if the \fB\-r\fP option is used.
.TP
.I "nan \- integer"
This is the number of records to process.  It is not read if the \fB\-r\fP
option is used.  A value of zero means all subsequent records in the file;
this is the default.
.TP
.I low_input \- string (converted to float)
Lower limit of range of input data to scale to the output range.  Data
below this value are clipped.  If the string "determine from file" is
entered, the limit is determined from the file.  This parameter is not
read if the
.B \-i
option is specified.  If the command-line option is not specified and
the parameter is not present in the parameter file, the default used is
the minimum value of the input data.  
.TP
.I high_input \- string (converted to float)
Upper limit of range of input data to scale to the output range.  Data
above this value are clipped.  If the string "determine from file" is
entered, the limit is determined form the file.	This parameter is not
read if the
.B \-i
option is specified.  If the command-line option is not specified and
the parameter is not present in the parameter file, the default used is
the maximum value of the input data.  
.TP
.I low_output \- float
Lower limit of range of output data.  Data in the input range are 
scaled to the output range.  This parameter is not read if the
.B \-o
option is specified.  If the command-line option is not specified and
the parameter is not present in the parameter file, the default used
is \-64.  
.TP
.I high_output \- float
Upper limit of range of output data.  Data in the input range are 
scaled to the output range.  This parameter is not read if the
.B \-o
option is specified.  If the command-line option is not specified and
the parameter is not present in the parameter file, the default used
is +50.  
.TP
.I "sf \- float"
Specifies the value to be used for the generic header item \fIsf\fP
in \fIoutput\fP.
As a result, if the frequency format of
.I output
is SYM_CTR, the frequency scale will range from 0
to \fIsf\fP/2 when \fIoutput\fP is displayed with \fIwaves\fP+.
(If the output
.I freq_format
is ARB_FIXED,
.I sf
plays no role in setting the frequency scale, which is determined
by an explicit list of values in the header instead\*-see the parameters
.I freq_format
and
.IR freqs. )
If the parameter value is 0 (the default), the generic's value is set to
2*(\fIsize\fP \- 1), where \fIsize\fP is the size of the input field
\fIfea_field\fP (this is done so that the "frequencies" correspond to
element number in the input field).  This parameter is not read if 
the option \-\fBs\fP is used.  
.PP
Remember that command line option values override parameter file
values.
.SH "ESPS COMMON"
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overridden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.PP
The following items are written into the ESPS Common file 
provided that \fIoutput\fP is not <stdout>.  
.IP
.I "start \- integer"
.IP
The starting point from the input file.
.sp
.I "nan \- integer"
.IP
The number of points in the selected range.
.sp
.I "prog \- string"
.IP
This is the name of the program ("tofspec" in this case).
.sp
.I "filename \- string"
.IP
The name of the input file \fIinput\fP.
.SH ESPS HEADERS
.PP
The \fIoutput\fP header is a FEA_SPEC file header.  The generic items
\fIstart\fP and \fInan\fP  are written to store the range of input
data records processed.  The items \fIlow_input, high_input,
low_output,\fP and \fIhigh_output\fP 
are written to record the range of input data that was transformed to
the output data.  If the \fB\-d\fP was used the log of the 
input range limits are also written as generics \fIlog_low_input\fP
and \fIlog_high_input\fP.  
.PP
The generic header item \fIstart_time\fP is written in the output
file.  The value written is computed by taking the \fIstart_time\fP
value from the header of the input file (or zero, if such a header
item doesn't exist) and adding to it the relative time from the first
record in the file to the first record processed.  The computation of
\fIstart_time\fP depends on the value of the generic header item
\fIrecord_freq\fP in the input file.  If this item is not present,
\fIstart_time\fP is just copied from the input file to the output
file.  
.PP
The generic header item \fIsf\fP is written in the output file;
the value is specified by the parameter \fIsf\fP or the value given in the
\fB\-s\fP option.
See the discussions of the parameter and the option for details.
.PP
If the input file is tagged, then the output, and the generic header item
.I src_sf
is written in the output file header.
In that case, the value of
.I src_sf
is copied from the input if the input header contains a
.I src_sf
item; otherwise the values is taken from
.I sf
in the input header.
(A warning message is printed if a tagged input file has neither
.I src_sf
nor
.I sf
in its header.)
.PP
As usual, the command line is added as a comment and the header of
\fIinput\fP is added as a source file to \fIoutput\fP.  
.SH "SEE ALSO"
.PP
.nf
\fIFEA_SPEC\fP(5\-\s-1ESPS\s+1), \fIxwaves\fP(1\-\s-1ESPS\s+1), \fIaddfea\fP(1\-\s-1ESPS\s+1),
\fIaddfeahd\fP(1\-\s-1ESPS\s+1), FEA(5\-\s-1ESPS\s+1), \fIimage\fP(1\-\s-1ESPS\s+1),
\fIplotsgram\fP(1\-\s-1ESPS\s+1)
.fi
.SH "WARNINGS AND DIAGNOSTICS"
.PP
.I tofspec
will exit with an error message if any of the following are true:
the \fB\-d\fP is used and any of the input data are not positive; 
.I input
does not exist or is not an ESPS FEA file; 
the data field does not exist in
.I input;
.PP
.SH "BUGS"
.PP
None known.  
.SH "AUTHOR"
.PP
Manual page and program by John Shore.  



ESPS/general/src/tofspec/tofspec.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1988-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1993  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  John Shore, Entropic Speech, Inc.
 * Checked by:
 * Revised by:
 *
 * This program takes the data from a given FEA field in an input file
 * and represents it as a "power spectrum" in an output FEA_SPEC
 * file.  The main purpose of this program is to allow waves+ users
 * to display arbitrary FEA data as spectrograms.
 *
 */

static char *sccs_id = "@(#)tofspec.c	1.12	9/5/95	ESI/ERL";

#define VERSION "1.12"
#define DATE "9/5/95"

#include <stdio.h>
#include <esps/esps.h>
#include <esps/unix.h>
#include <esps/fea.h>
#include <esps/feaspec.h>

#define ERROR(text) { \
  (void) fprintf(stderr, "%s: %s - exiting\n", ProgName, (text)); \
  exit(1);}
#define REQUIRE(test, text) {if (!(test)) ERROR(text)}

#define SYNTAX \
USAGE("tofspec [-d] [-f fea_field] [-i input_range] [-o output_range]\n\t[-r record_range] [-s sf] [-v freqs] [-x debug_level] [-F freq_format]\n\t[-P params] [-R] [-S] input output");

/* default upper and lower limits; only use 128-13 values 
   since top 13 levels of color map are used for things other than
   data values in waves+*/

#define HIGH_OUT 50
#define LOW_OUT -64.0

double          log10();
double          getsym_d();
char		*get_cmd_line();
void		lrange_switch();
void		copy_fea_fields();
void		copy_fea_rec(), frange_switch();
char            *type_convert();

static float	*flist_switch();


char	*ProgName = "tofspec";
char	*Version = VERSION;
char	*Date = DATE;
int	debug_level = 0;	/* debug level, global for library*/

/*
 * MAIN PROGRAM
 */

main(argc, argv)
    int  argc;
    char **argv;
{
    extern int	    optind;	/* for use of getopt() */
    extern char	    *optarg;	/* for use of getopt() */
    int		    ch;		/* command-line option letter */
    char	    *field_name; /* name of input field to transform */
    int		    field_type;	/* data type of input field */
    int		    dflag = NO;	/* -d option specified? */
    int             fflag = NO; /* -f option specified? */
    int 	    iflag = NO;	/* -l option specified? */
    int 	    oflag = NO;	/* -o option specified? */
    int		    rflag = NO;	/* -r option specified? */
    int             sflag = NO; /* -s option specified? */
    int             vflag = NO; /* -v option specified? */
    int             Fflag = NO; /* -F option specified? */
    int 	    Rflag = NO;	/* -R option specified? */
    int 	    Sflag = NO;	/* -S option specified? */
    int  determine_limits = NO; /* true if display limits from file */
    char	    *rec_range;	/* arguments of -r option */
    char            *in_range;	/* data range for full output range */
    char            *out_range;	/* output range */
    char	    *fmt_name;	/* string naming freq_format */
    int		    freq_format; /* FEA_SPEC frequenc format:
				    SYM_EDGE, ARB_FIXED, etc. */
    char	    *freq_list;	/* string giving frequeny values */
    float	    *freqs = NULL; /* frequency values */
    long	    num_freqs;	/* number of frequency values */
    long	    start_rec;	/* starting record number */
    long	    end_rec;	/* ending record number */
    long	    num_recs;	/* number of records to read
				   (0 means all up to end of file) */
    long	    num_read;	/* number of records actually read */

    char	    *param_name = NULL; /* parameter file name */

    char	    *iname;	/* input file name */
    FILE	    *ifile;	/* input stream */
    struct header   *ihd;	/* input file header */
    struct fea_data *irec;	/* input record */
    char	    *f_ptr;	/* pointer (untyped) to data
				   field in input record */
    long	    size;	/* size of input data field */
    short           rank = 0;	/* rank of input data field */
    float	    *data;	/* input data as doubles */
    float           dtemp;	/* temp location for one input value */
    char	    *oname;	/* output file name */
    FILE	    *ofile;	/* output stream */
    struct header   *ohd;	/* output file header */
    struct feaspec  *orec;	/* output record */
    double          low_in = FLT_MAX; /*input range limits*/
    double          high_in = -FLT_MAX; 
    double          log_low_in, log_high_in; /*log input limits*/
    double	    high_out = HIGH_OUT; /*output range limits*/
    double          low_out = LOW_OUT;
    char            *high_in_s = NULL; /*strings for param file input*/
    char            *low_in_s = NULL; 

    double  datamin, datamax;	/* max and min data values */
    int		    j,k;	/* loop indices */
    double 	    sf;		/* sampling frequency */
    float   	    scale;	/* scaling constant */

    /*
     * Parse command-line options.
     */

    while ((ch = getopt(argc, argv, "df:i:o:r:s:v:x:F:P:RS")) != EOF)
        switch (ch)
	{
	case 'd':
	    dflag = YES;
	    break;
	case 'f':
	    field_name = optarg;
	    fflag++;
	    break;
	case 'i':
	    iflag = YES;
	    in_range = optarg;
	    break;
	case 'o':
	    oflag = YES;
	    out_range = optarg;
	    break;
	case 'r':
	    rflag = YES;
	    rec_range = optarg;
	    break;
	case 's':
	    sflag = YES;
	    sf = atof(optarg);
	    break;
	case 'v':
	    vflag = YES;
	    freq_list = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'F':
	    Fflag = YES;
	    fmt_name = optarg;
	    break;
	case 'P':
	    param_name = optarg;
	    break;
	case 'R':
	    Rflag = YES;
	    break;
	case 'S':
	    Sflag = YES;
	    break;
	default:
	    SYNTAX;
	    break;
	}

    if (debug_level && dflag) 
	Fprintf(stderr, "tofspec: -d option specified, will take logs\n");

    if (debug_level && iflag) 
	Fprintf(stderr, "tofspec:  limits determined from -i option\n");

    if (iflag && Sflag) 
	Fprintf(stderr, "tofspec: -i option ignored since -S option given\n");

    /*
     * Process file names and open files.
     */

    if (argc - optind > 2)
    {
	Fprintf(stderr,
		"%s: too many file names specified.\n", ProgName);
	SYNTAX;
    }
    if (argc - optind < 2)
    {
	Fprintf(stderr,
		"%s: too few file names specified.\n", ProgName);
	SYNTAX;
    }

    iname = eopen(ProgName,
		  argv[optind++], "r", FT_FEA, NONE, &ihd, &ifile);

    oname = eopen(ProgName,
		  argv[optind++], "w", NONE, NONE, &ohd, &ofile);

    if (debug_level)
	Fprintf(stderr, "tofspec: Input file: %s\nOutput file: %s\n",
		iname, oname);

    /*
     * Get parameter values.
     */

    if (ifile != stdin)
	(void) read_params(param_name, SC_CHECK_FILE, iname);
    else
	(void) read_params(param_name, SC_NOCOMMON, iname);

    /* input range */

    if (iflag)
	frange_switch(in_range, &low_in, &high_in);

    else if (!Sflag) {
	if (symtype("low_input") != ST_UNDEF) {
	    low_in_s = getsym_s("low_input");
	    if (strcmp(low_in_s, "determine from file") != 0)
		low_in = atof(low_in_s);
	}
	if (symtype("high_input") != ST_UNDEF) {
	    high_in_s = getsym_s("high_input");
	    if (strcmp(high_in_s, "determine from file") != 0)
		high_in = atof(high_in_s);
	}
    }

    determine_limits = (low_in == FLT_MAX || high_in == -FLT_MAX);

    if (debug_level) 
	Fprintf(stderr, "tofspec: determine_limits = %d\n", determine_limits);

    if (determine_limits && (strcmp(iname, "<stdin>") == 0)) {
	Fprintf(stderr, 
		"tofspec: can't use stdin when input range determined from file\n");
	exit(1);
    }

    /* output limits */

    if (oflag)
	frange_switch(out_range, &low_out, &high_out);
    else
    {
	if (symtype("low_output") != ST_UNDEF) 
	    low_out = getsym_d("low_output");
	if (symtype("high_output") != ST_UNDEF) 
	    high_out = getsym_d("high_output");
    }

    if (low_out < -64.0) {
	Fprintf(stderr, 
		"tofspec: WARNING - output data type is BYTE, so low_out limited to -64.0\n");
	low_out = -64.0;
    }

    if (high_out > 63.0) {
	Fprintf(stderr, 
		"tofspec: WARNING - output data type is BYTE, so high_out limited to 63.0\n");
	high_out = 63.0;
    }

    if (debug_level)
	Fprintf(stderr, "tofspec: output range is %g, %g.\n", 
		low_out, high_out);    

    /* field name */

    if (!fflag)
	field_name =
	    (symtype("field_name") != ST_UNDEF)
		? getsym_s("field_name")
		    : "sd";

    if (debug_level)
	Fprintf(stderr,
		"  field_name is \"%s\".\n", field_name);

    /* range of input records */

    if (rflag)
    {
	start_rec = 1;
	end_rec = LONG_MAX;
	lrange_switch(rec_range, &start_rec, &end_rec, 0);
	num_recs =
	    (end_rec != LONG_MAX)
		? end_rec - start_rec + 1
		    : 0;
    }
    else
    {
	start_rec =
	    (symtype("start") != ST_UNDEF)
		? getsym_i("start")
		    : 1;
	num_recs =
	    (symtype("nan") != ST_UNDEF)
		? getsym_i("nan")
		    : 0;
	end_rec = (num_recs != 0)
	    ? start_rec - 1 + num_recs
		: LONG_MAX;
    }

    if (debug_level)
	Fprintf(stderr, "start_rec: %ld.  end_rec: %ld.  num_recs: %ld.\n",
		start_rec, end_rec, num_recs);

    REQUIRE(start_rec >= 1, "can't start before beginning of file");
    REQUIRE(end_rec >= start_rec, "empty range of records specified");

    /*
     * Get info on input field
     */

    field_type = get_fea_type(field_name, ihd);

    REQUIRE(field_type != UNDEF, "specified field not in input file");

    size = get_fea_siz(field_name, ihd, &rank, (long **) NULL);

    if (rank != 1) 
	Fprintf(stderr, 
		"tofspec: WARNING, field %s not a vector.\n", field_name);

    REQUIRE(size != 0,
	    "size == 0, bad data field definition in input file");

    /* sf */
    /* If -s is used, we have a forced setting for sf; otherwise
       we fake up as element numbers. */

    if (!sflag)
	sf =
	    (symtype("sf") != ST_UNDEF)
		? getsym_d("sf")
		    : 0;	/* means fake up element numbers as frequencies*/

    if (sf == 0) 
	sf = 2*(size - 1);	/* fake up element numbers as frequencies*/

    /* freq_format */

    if (!Fflag)
    {
	fmt_name =
	    (symtype("freq_format") != ST_UNDEF)
		? getsym_s("freq_format")
		    : "SYM_EDGE";
	REQUIRE(fmt_name, "Parameter \"freq_format\" not STRING");
    }

    switch (freq_format = lin_search(spfmt_names, fmt_name))
    {
    case SPFMT_ARB_FIXED:
	if (vflag)
	{
	    freqs = flist_switch(freq_list, &num_freqs);
	    REQUIRE(num_freqs == size,
		    "wrong number of specified frequencies");
	}
	else
	{
	    REQUIRE(symtype("freqs") == ST_FARRAY,
		    "Paramter \"freqs\" undefined or not float array");
	    num_freqs = symsize("freqs");
	    REQUIRE(num_freqs == size,
		    "wrong number of specified frequencies");
	    freqs = (float *) malloc(sizeof(float) * size);
	    REQUIRE(freqs, "couldn't allocate storage for freqs array");
	    (void) getsym_fa("freqs", freqs, (int) num_freqs);
	}

	break;
    case SPFMT_SYM_EDGE:
	break;
    case SPFMT_ASYM_EDGE:
    case SPFMT_SYM_CTR:
    case SPFMT_ASYM_CTR:
    case SPFMT_ARB_VAR:
	ERROR("freq format not yet supported");
	break;
    default:
	ERROR("freq format not recognized");
	break;
    }

    /* possible loop to determine data limits*/

    irec = allo_fea_rec(ihd);
    f_ptr = get_fea_ptr(irec, field_name, ihd);

    REQUIRE(f_ptr, "NULL pointer returned from get_fea_ptr");

    data = calloc_f((unsigned) size);

    REQUIRE(data,
	    "can't allocate space for array of data as doubles");

    if (determine_limits) {

	if (debug_level) 
	    Fprintf(stderr, "tofspec: computing display limits from data\n");
      
	datamin = FLT_MAX;
	datamax = -FLT_MAX;
      
	num_read = start_rec - 1,
	fea_skiprec(ifile, num_read, ihd);
      
	for ( ;
	     num_read < end_rec && get_fea_rec(irec, ihd, ifile) != EOF;
	     num_read++
	     )
	{
	    if (debug_level > 2)
		Fprintf(stderr, 
			"Record number %ld read.\n", num_read + 1);
	  
	    (void) type_convert(size, f_ptr, field_type, 
				(char *) data, FLOAT, (void (*)()) NULL);

	    for (j = 0; j < size; j++) {
		if (data[j] < datamin) datamin = data[j];
		if (data[j] > datamax) datamax = data[j];
	    }
	}
      
	if (num_read < end_rec && num_recs != 0)
	    Fprintf(stderr, "tofspec: Only %ld records read.\n", 
		    num_read - start_rec + 1);

	if (low_in == FLT_MAX) low_in = datamin;
	if (high_in == -FLT_MAX) high_in = datamax;

	rewind(ifile);
	ihd = read_header(ifile);
    }

    if (debug_level)
	Fprintf(stderr, "tofspec: input range is %g, %g.\n", 
		low_in, high_in);

    if (dflag) {
	if (low_in <= 0) {
	    Fprintf(stderr, 
		    "tofspec: can't use -d option with negative or zero data\n");
	    exit(1);
	}
	else {
	    log_low_in = log10(low_in);
	    log_high_in = log10(high_in);
	    if (debug_level) 
		Fprintf(stderr, 
			"tofspec: log limits on data values: %g, %g.\n", 
			log_low_in, log_high_in);
	}
    }

    /*set scale values for data */

    if (dflag) 
	scale = (high_out - low_out) / (log_high_in - log_low_in);
    else 
	scale = (high_out - low_out) / (high_in - low_in);

    if (debug_level) Fprintf(stderr, "tofspec: scale = %g\n", scale);

    /*
     * Create and write output-file header
     */

    ohd = new_header(FT_FEA);

    if (init_feaspec_hd(ohd, NO, freq_format, SPTYP_DB,
			YES, size, SPFRM_NONE, freqs, sf, (long) 0,
			BYTE) != 0) {
	Fprintf(stderr, "tofspec: Error filling FEA_SPEC header\n");
	exit(1);
    }

    ohd->common.tag = ihd->common.tag;
    if (ihd->common.tag
	&& !copy_genhd(ohd, ihd, "src_sf"))
    {
	if (genhd_type("sf", (int *) NULL, ihd) == ST_UNDEF)
	    Fprintf(stderr,
		    "tofspec: no \"src_sf\" or \"sf\" in tagged file.\n");
	else
	    *add_genhd_f("src_sf", (float *) NULL, 1, ohd) =
		get_genhd_val("sf", ihd, 1.0);
    }

    ohd->variable.refer = ihd->variable.refer;
    ohd->variable.refhd = ihd->variable.refhd;

    (void) strcpy(ohd->common.prog, ProgName);
    (void) strcpy(ohd->common.vers, Version);
    (void) strcpy(ohd->common.progdate, Date);

    add_source_file(ohd, iname, ihd);
    add_comment(ohd, get_cmd_line(argc, argv));

    *add_genhd_l("nan", (long *) NULL, 1, ohd) = num_recs;
    *add_genhd_l("start", (long *) NULL, 1, ohd) = start_rec;
    *add_genhd_f("low_input", (float *) NULL, 1, ohd) = low_in;
    *add_genhd_f("high_input", (float *) NULL, 1, ohd) = high_in;
    *add_genhd_f("low_output", (float *) NULL, 1, ohd) = low_out;
    *add_genhd_f("high_output", (float *) NULL, 1, ohd) = high_out;

    if (!Sflag) 
	*add_genhd_f("scale_factor", (float *) NULL, 1, ohd) = scale;

    if (dflag) {
	*add_genhd_f("log_low_input", (float *) NULL, 1, ohd) = log_low_in;
	*add_genhd_f("log_high_input", (float *) NULL, 1, ohd) = log_high_in;
    }

    update_waves_gen(ihd, ohd, (float)start_rec, 1.0);

    if (debug_level)
	Fprintf(stderr, "tofspec: Writing output header to file.\n");

    write_header(ohd, ofile);

    orec = allo_feaspec_rec(ohd, FLOAT);

    /*
     * Main read-write loop
     */

    /* now read/write loop for putting out FEA_SPEC file*/

    num_read = start_rec - 1,
    fea_skiprec(ifile, num_read, ihd);

    for ( ;
	 num_read < end_rec && get_fea_rec(irec, ihd, ifile) != EOF;
	 num_read++
	 )
    {
	if (debug_level > 1)
	    Fprintf(stderr, "Record number %ld read.\n", num_read + 1);

	(void) type_convert(size, f_ptr, field_type, 
			    (char *) data, FLOAT, (void (*)()) NULL);

	for (j = 0; j < size; j++) {

	    if (Rflag)		/*reverse elements*/
		k = size - 1 - j;
	    else		/*element order unchanged*/
		k = j;

	    if (Sflag) {	/*no scaling*/
		if (dflag) 
		    dtemp = log10(data[k]);
		else
		    dtemp = data[k];
		if (dtemp < low_out)
		    orec->re_spec_val[j] = low_out;
		else if (dtemp > high_out)
		    orec->re_spec_val[j] = high_out;
		else 
		    orec->re_spec_val[j] = dtemp;
	    }
	    else {		/*scale the data to fit*/
		if (data[k] <= low_in)
		    orec->re_spec_val[j] = low_out;
		else if (data[k] >= high_in)
		    orec->re_spec_val[j] = high_out;
		else {
		    if (dflag)
			orec->re_spec_val[j] = 
			    low_out + scale * (log10(data[k]) - log_low_in);
		    else
			orec->re_spec_val[j] = 
			    low_out + scale * (data[k] - low_in);
		}
	    }
	}
	
	if (ohd->common.tag == YES)
	    *orec->tag = irec->tag;

	(void) put_feaspec_rec(orec, ohd, ofile);
    }

    if (num_read < end_rec && num_recs != 0)
	Fprintf(stderr, "tofspec: Only %ld records read.\n", 
		num_read - start_rec + 1);
 
    /*
     * Write common
     */

    if (strcmp(oname, "<stdout>") != 0 && strcmp(iname, "<stdin>") != 0) {
	(void) putsym_s("filename", iname);
	(void) putsym_s("prog", ProgName);
	(void) putsym_i("start", (int) start_rec);
	(void) putsym_i("nan", (int) num_recs);
    }

    exit(0);
    /*NOTREACHED*/
}


#define INIT_SIZE   200
#define DELIM       ", "

char	*savestring();

static float *
flist_switch(text, len)
    char    *text;
    long    *len;
{
    char    *list, *token;
    long    alloc_len;
    float   *vals;
    long    num;

    spsassert(text, "flist_switch: NULL text");
    list = savestring(text);
    alloc_len = INIT_SIZE;
    vals = (float *) malloc(sizeof(float) * alloc_len);
    if (!vals)
	return NULL;
    for (token = strtok(list, DELIM), num = 0;
	 token;
	 token = strtok((char *) NULL, DELIM), num++)
    {
	if (num == alloc_len)
	{
	    alloc_len += alloc_len/2;
	    vals =
		(float *) realloc((char *) vals, sizeof(float) * alloc_len);
	    if (!vals)
		return NULL;
	}

	vals[num] = atof(token);
    }

    free((char *) list);
    if (len)
	*len = num;
    return (float *) realloc((char *) vals, sizeof(float) * num);
}
